internal/c/parts/gui/gui.cpp

```diff
@@ -0,0 +1,365 @@
+        qbs_set(aMessage, qbs_add(qbsMessage, qbs_new_txt_len("\0", 1)));
+    else
+        qbs_set(aMessage, qbs_new_txt_len("\0", 1));
+
+    if (passed & 4)
+        qbs_set(aIconType, qbs_add(qbsIconType, qbs_new_txt_len("\0", 1)));
+    else
+        qbs_set(aIconType, qbs_new_txt("info")); // info if not passed
+
+    tinyfd_messageBox((const char *)aTitle->chr, (const char *)aMessage->chr, "ok", (const char *)aIconType->chr, 1);
+}
+
+/// @brief Shows an input box for getting a string from the user
+/// @param qbsTitle Title of the dialog box
+/// @param qbsMessage The message or prompt that will be displayed
+/// @param qbsDefaultInput [OPTIONAL] The default response that can be changed by the user
+/// @param passed How many parameters were passed?
+/// @return The user response or an empty string if the user cancelled
+qbs *func__guiInputBox(qbs *qbsTitle, qbs *qbsMessage, qbs *qbsDefaultInput, int32_t passed) {
+    static qbs *aTitle = nullptr;
+    static qbs *aMessage = nullptr;
+    static qbs *aDefaultInput = nullptr;
+    char *sDefaultInput = nullptr;
+
+    if (!aTitle)
+        aTitle = qbs_new(0, 0);
+
+    if (!aMessage)
+        aMessage = qbs_new(0, 0);
+
+    if (!aDefaultInput)
+        aDefaultInput = qbs_new(0, 0);
+
+    qbs_set(aTitle, qbs_add(qbsTitle, qbs_new_txt_len("\0", 1)));
+    qbs_set(aMessage, qbs_add(qbsMessage, qbs_new_txt_len("\0", 1)));
+
+    if (passed) {
+        qbs_set(aDefaultInput, qbs_add(qbsDefaultInput, qbs_new_txt_len("\0", 1)));
+        sDefaultInput =
+            aDefaultInput->len == 1 ? nullptr : (char *)aDefaultInput->chr; // if string is "" then password box, else we pass the default input as is
+    } else {
+        qbs_set(aDefaultInput, qbs_new_txt_len("\0", 1)); // input box by default
+        sDefaultInput = (char *)aDefaultInput->chr;
+    }
+
+    auto sInput = tinyfd_inputBox((const char *)aTitle->chr, (const char *)aMessage->chr, (const char *)sDefaultInput);
+
+    return qbs_new_txt(sInput);
+}
+
+/// @brief Shows the browse for folder dialog box
+/// @param qbsTitle Title of the dialog box
+/// @param qbsDefaultPath [OPTIONAL] The default path from where to start browsing
+/// @param passed How many parameters were passed?
+/// @return The path selected by the user or an empty string if the user cancelled
+qbs *func__guiSelectFolderDialog(qbs *qbsTitle, qbs *qbsDefaultPath, int32_t passed) {
+    static qbs *aTitle = nullptr;
+    static qbs *aDefaultPath = nullptr;
+
+    if (!aTitle)
+        aTitle = qbs_new(0, 0);
+
+    if (!aDefaultPath)
+        aDefaultPath = qbs_new(0, 0);
+
+    qbs_set(aTitle, qbs_add(qbsTitle, qbs_new_txt_len("\0", 1)));
+
+    if (passed)
+        qbs_set(aDefaultPath, qbs_add(qbsDefaultPath, qbs_new_txt_len("\0", 1)));
+    else
+        qbs_set(aDefaultPath, qbs_new_txt_len("\0", 1));
+
+    auto sFolder = tinyfd_selectFolderDialog((const char *)aTitle->chr, (const char *)aDefaultPath->chr);
+
+    return qbs_new_txt(sFolder);
```

Comment:
This seems like it could be a potential problem. I assume `sFolder` is backed by a static buffer in `tinyfd`. We probably want to duplicate it here so that the returned `qbs` doesn't randomly change if the user calls `_SelectFolderDialog` a second time. This goes for some of the other functions as well.

Admittedly I don't understand `qbs` well enough to know exactly how to do this. It looks like other functions that return strings use `qbs_new()` to create a new string and then copy the contents in, presumably that would work here too 🤷

Reply:
That is a good catch. I was assuming that `qbs_new_txt` and `qbs_new_txt_len` makes copies of the strings. I looked at those closely and it seems those aren't. They simply do a `newstr->chr = (uint8 *)txt`. `func_string` does what you are saying.  I'll rewrite this and the others.

Reply:
Yeah I noticed the behavior of `qbs_new_txt` a little while back, it seems designed for creating `qbs`s out of static strings, which I suppose makes sense. As part of the `qbs` refactor ideally we'll document some of these functions 😆...

Reply:
This is done. I modified the functions to return a `qbs` like `func_string` does.